*** blocks/course_overview/block_course_overview.php	2011-01-19 16:01:39.000000000 -0800
--- blocks/course_overview/block_course_overview.php	2012-11-14 14:28:25.000000000 -0800
***************
*** 49,54 ****
--- 49,56 ----
          $this->content = new stdClass();
          $this->content->text = '';
          $this->content->footer = '';
+ 	$visiblity_filter = $CFG->block_course_overview_visible;
+ 	$start_date_filter = $CFG->block_course_overview_start_date;
  
          $content = array();
  
***************
*** 64,70 ****
              $courses_limit = $courses_limit + 1;
          }
  
!         $courses = enrol_get_my_courses('id, shortname, modinfo', 'visible DESC,sortorder ASC', $courses_limit);
          $site = get_site();
          $course = $site; //just in case we need the old global $course hack
  
--- 66,79 ----
              $courses_limit = $courses_limit + 1;
          }
  
! 	$sortorder = 'sortorder ASC';
!         if($start_date_filter){
! 	  $sortorder = 'startdate DESC';
! 	}
! 	if($visiblity_filter){
! 	  $sortorder = 'visible DESC, ' . $sortorder;
! 	}
!         $courses = enrol_get_my_courses('id, shortname, modinfo', $sortorder , $courses_limit);
          $site = get_site();
          $course = $site; //just in case we need the old global $course hack
  
*** blocks/course_overview/lang/en/block_course_overview.php	2010-05-04 06:04:36.000000000 -0700
--- blocks/course_overview/lang/en/block_course_overview.php	2012-10-31 14:05:28.000000000 -0700
***************
*** 1,2 ****
--- 1,7 ----
  <?php
  $string['pluginname'] = 'Course overview';
+ $string['visible'] = 'Sort by visiblity';
+ $string['visible_desc'] = 'Select this option, to group courses by visibility.';
+ $string['start_date'] = 'Sort by course start date' ;
+ $string['start_date_desc'] = 'Select this option, to sort courses by start date on student and instructor\'s My Home page. Default is course sort order.';
+ 
