diff -Naur forum/backup/moodle1/lib.php forum/backup/moodle1/lib.php
--- forum/backup/moodle1/lib.php	2012-11-19 10:54:38.000000000 -0800
+++ forum/backup/moodle1/lib.php	2012-11-19 14:32:43.000000000 -0800
@@ -73,6 +73,11 @@
      * Converts /MOODLE_BACKUP/COURSE/MODULES/MOD/FORUM data
      */
     public function process_forum($data) {
+        //rru-specific start
+        // required for fix below
+        global $CFG;
+        //rru-specific end
+        
         // get the course module id and context id
         $instanceid     = $data['id'];
         $cminfo         = $this->get_cminfo($instanceid);
@@ -86,6 +91,15 @@
         $this->fileman->filearea = 'intro';
         $this->fileman->itemid   = 0;
         $data['intro'] = moodle1_converter::migrate_referenced_files($data['intro'], $this->fileman);
+        
+        //rru-specific start 
+        // Set the introformat to FORMAT_HTML to fix RRU Moodle Bug 161 (disappearing editor)
+        // if an additional editor is enabled
+        if ($CFG->texteditors !== 'textarea') {
+            $data['intro']       = text_to_html($data['intro'], false, false, true);
+            $data['introformat'] = FORMAT_HTML;
+        }
+        //rru-specific end
 
         // start writing forum.xml
         $this->open_xml_writer("activities/forum_{$this->moduleid}/forum.xml");
diff -Naur forum/index.php forum/index.php
--- forum/index.php	2012-11-19 10:54:38.000000000 -0800
+++ forum/index.php	2012-11-19 14:32:43.000000000 -0800
@@ -69,6 +69,10 @@
 $strno           = get_string('no');
 $strrss          = get_string('rss');
 
+//rru-specific start
+$strdraftstitle  = get_string('draftsheader', 'local_draftposts');
+//rru-specific end
+
 $searchform = forum_search_form($course);
 
 
@@ -78,6 +82,12 @@
 $generaltable->head  = array ($strforum, $strdescription, $strdiscussions);
 $generaltable->align = array ('left', 'left', 'center');
 
+//rru-specific start
+// Support draftposts customization - include 'Drafts' header
+$generaltable->head[] = $strdraftstitle;
+$generaltable->align[] = 'center';
+//rru-specific start
+
 if ($usetracking = forum_tp_can_track_forums()) {
     $untracked = forum_tp_get_untracked_forums($USER->id, $course->id);
 
@@ -239,7 +249,14 @@
         $forumlink = "<a href=\"view.php?f=$forum->id\" $style>".format_string($forum->name,true)."</a>";
         $discussionlink = "<a href=\"view.php?f=$forum->id\" $style>".$count."</a>";
 
-        $row = array ($forumlink, $forum->intro, $discussionlink);
+        //rru-specific start
+        // Support Draftposts customization - show number of drafts for each forum for the current user
+        $draftcount = $DB->count_records('rrudraft_forum_posts', array('userid'=>$USER->id, 'forumid'=>$forum->id));
+        $dpurl = new moodle_url('/local/draftposts/draftposts.php', array('f'=>$forum->id));
+        $draftlink = "<a href=\"$dpurl\" $style>$draftcount</a>";
+
+        $row = array ($forumlink, $forum->intro, $discussionlink, $draftlink);
+        //rru-specific end
         if ($usetracking) {
             $row[] = $unreadlink;
             $row[] = $trackedlink;    // Tracking.
@@ -281,6 +298,12 @@
 $learningtable->head  = array ($strforum, $strdescription, $strdiscussions);
 $learningtable->align = array ('left', 'left', 'center');
 
+//rru-specific start
+// Support draftposts customization - include 'Drafts' header
+$learningtable->head[] = $strdraftstitle;
+$learningtable->align[] = 'center';
+//rru-specific end
+
 if ($usetracking) {
     $learningtable->head[] = $strunreadposts;
     $learningtable->align[] = 'center';
@@ -371,7 +394,15 @@
             $forumlink = "<a href=\"view.php?f=$forum->id\" $style>".format_string($forum->name,true)."</a>";
             $discussionlink = "<a href=\"view.php?f=$forum->id\" $style>".$count."</a>";
 
-            $row = array ($printsection, $forumlink, $forum->intro, $discussionlink);
+            //rru-specific start
+            // Support Draftposts customization - show number of drafts for each forum for the current user
+            $draftcount = $DB->count_records('rrudraft_forum_posts', array('userid'=>$USER->id, 'forumid'=>$forum->id));
+            $dpurl = new moodle_url('/local/draftposts/draftposts.php', array('f'=>$forum->id));
+            $draftlink = "<a href=\"$dpurl\" $style>$draftcount</a>";
+            //$draftlink = "<a href=\"$CFG->dirroot\local\draftposts\draftposts.php?f=$forum->id\" $style>$draftcount</a>";
+
+            $row = array ($printsection, $forumlink, $forum->intro, $discussionlink, $draftlink);
+            //rru-specific end
             if ($usetracking) {
                 $row[] = $unreadlink;
                 $row[] = $trackedlink;    // Tracking.
diff -Naur forum/lib.php forum/lib.php
--- forum/lib.php	2012-11-19 10:54:38.000000000 -0800
+++ forum/lib.php	2012-11-19 14:32:43.000000000 -0800
@@ -5297,6 +5297,14 @@
     $getuserlastmodified = ($displayformat == 'header');
 
     if (! $discussions = forum_get_discussions($cm, $sort, $fullpost, null, $maxdiscussions, $getuserlastmodified, $page, $perpage) ) {
+        //rru-specific start
+        // Support draftposts customization - don't call if the forum is on the front page
+        if ($cm->course != 1) {
+            require_once('../../local/draftposts/lib.php');
+            dp_print_drafts_link($forum->id);
+        }
+        //rru-specific end 
+        
         echo '<div class="forumnodiscuss">';
         if ($forum->type == 'news') {
             echo '('.get_string('nonews', 'forum').')';
@@ -5331,6 +5339,12 @@
         }
     }
 
+    //rru-specific start
+    // Support draftposts customization
+    require_once('../../local/draftposts/lib.php');
+    dp_print_drafts_link($forum->id);
+    //rru-specific end 
+ 
     $canviewparticipants = has_capability('moodle/course:viewparticipants',$context);
 
     $strdatestring = get_string('strftimerecentfull');
@@ -5554,7 +5568,13 @@
     $post->subject = format_string($post->subject);
 
     $postread = !empty($post->postread);
-
+    
+    //rru-specific start
+    // Support draftposts customization
+    require_once('../../local/draftposts/lib.php');
+    dp_print_drafts_link($forum->id, $discussion->id);
+    //rru-specific end 
+    
     forum_print_post($post, $discussion, $forum, $cm, $course, $ownpost, $reply, false,
                          '', '', $postread, true, $forumtracked);
 
diff -Naur forum/post_form.php forum/post_form.php
--- forum/post_form.php	2012-11-19 10:54:38.000000000 -0800
+++ forum/post_form.php	2012-11-19 14:32:43.000000000 -0800
@@ -141,6 +141,13 @@
         }
         $this->add_action_buttons(false, $submit_string);
 
+        //rru-specific start
+        //To support RRU Draft Posts
+        //Important: call needs to be after $this->add_action_buttons() function
+        require_once('../../local/draftposts/lib.php');
+        dp_return_save_draft_button($mform);
+        //rru-specific end
+
         $mform->addElement('hidden', 'course');
         $mform->setType('course', PARAM_INT);
 
diff -Naur forum/post.php forum/post.php
--- forum/post.php	2012-11-19 10:54:38.000000000 -0800
+++ forum/post.php	2012-11-19 14:32:43.000000000 -0800
@@ -331,6 +331,16 @@
                     notice("Sorry, but you are not allowed to delete that discussion!",
                             forum_go_back_to("discuss.php?d=$post->discussion"));
                 }
+                //rru-specific start
+                // Support draft post customization - delete drafts when a discussion thread is deleted
+                // Needs to delete drafts prior to deleting discussion so we can confirm user has permission
+                // Txn would be best, but such a rare instance, this will have to do for now
+                require_once('../../local/draftposts/lib.php');
+                if (!dp_delete_thread_drafts($post->forum, $discussion->groupid ? $discussion->groupid : 0, $post->discussion)) {
+                    echo $OUTPUT->notification(get_string('deletefail', 'local_draftposts'));
+                }
+                //rru-specific end
+                                
                 forum_delete_discussion($discussion, false, $course, $cm, $forum);
 
                 add_to_log($discussion->course, "forum", "delete discussion",
@@ -661,6 +671,14 @@
         add_to_log($course->id, "forum", "update post",
                 "$discussionurl&amp;parent=$fromform->id", "$fromform->id", $cm->id);
 
+        //rru-specific start
+        // Support draft post customization - delete draft when an existing post is saved
+        require_once('../../local/draftposts/lib.php');
+        if (!dp_delete_draft(null, $updatepost->forum, $groupid == -1 ? 0 : $groupid, $updatepost->discussion, $updatepost->parent)) {
+            echo $OUTPUT->notification(get_string('deletefail', 'local_draftposts'));
+        }
+        //rru-specific end
+        
         redirect(forum_go_back_to("$discussionurl"), $message.$subscribemessage, $timemessage);
 
         exit;
@@ -708,6 +726,14 @@
                 $completion->update_state($cm,COMPLETION_COMPLETE);
             }
 
+            //rru-specific start
+            // Support draft post customization - delete draft when new post is saved
+            require_once('../../local/draftposts/lib.php');
+            if (!dp_delete_draft(null, $addpost->forum, $groupid == -1 ? 0 : $groupid, $addpost->discussion, $addpost->parent)) {
+                echo $OUTPUT->notification(get_string('deletefail', 'local_draftposts'));
+            }
+            //rru-specific end
+                        
             redirect(forum_go_back_to("$discussionurl#p$fromform->id"), $message.$subscribemessage, $timemessage);
 
         } else {
@@ -769,6 +795,14 @@
                 $completion->update_state($cm,COMPLETION_COMPLETE);
             }
 
+            //rru-specific start
+            // Support draft post customization - delete draft when discussion is saved
+            require_once('../../local/draftposts/lib.php');
+            if (!dp_delete_draft(null, $discussion->forum, $discussion->groupid == -1 ? 0 : $discussion->groupid)) {
+                echo $OUTPUT->notification(get_string('deletefail', 'local_draftposts'));
+            }
+            //rru-specific end
+                        
             redirect(forum_go_back_to("view.php?f=$fromform->forum"), $message.$subscribemessage, $timemessage);
 
         } else {
@@ -874,5 +908,48 @@
 
 $mform_post->display();
 
+//rru-specific start
+// Support draftposts customization
+// NOTE: this needs to be the last thing done before the footer is displayed
+$module = array(
+    'name' => 'local_draftposts',
+    'fullpath' => '/local/draftposts/module.js',
+    'requires' => array('base', 'io', 'node', 'json'),
+    'strings' => array(
+        array('confirmload', 'local_draftposts')
+    )
+);
+
+$sqlconditions = array(
+    'forumid' => ($post->forum ? $post->forum : 0),
+    'discussionid' => ($post->discussion ? $post->discussion : 0),
+    'parentid' => ($post->parent ? $post->parent : 0),
+    'groupid' => ($post->groupid ? $post->groupid : 0),
+    'userid' => ($USER->id ? $USER->id : 0)  // use logged in user not post user for security purposes
+);
+
+$jscall = 'M.local_draftposts.init';
+$jsargs['cfg']['saveinterval'] = get_config('local_draftposts','saveinterval');
+$jsargs['cfg']['sesskey'] = sesskey();
+// Currently only support HTML format for editor
+$jsargs['cfg']['supportedformat'] = FORMAT_HTML;
+$jsargs['loaddraft'] = 0;
+
+// If we find a draft for this post - ask user if they want to load it
+if ($DB->get_field('rrudraft_forum_posts', 'id', $sqlconditions)) {
+    $jsargs['loaddraft'] = 1;
+}
+
+$PAGE->requires->js_module($module);
+$PAGE->requires->js_init_call($jscall, $jsargs, true, $module);
+
+// Add some info to the session for use in ajax_draftposts.php
+$SESSION->draftpost->forumid = $sqlconditions['forumid'];
+$SESSION->draftpost->discussionid = $sqlconditions['discussionid'];
+$SESSION->draftpost->parentid = $sqlconditions['parentid'];
+$SESSION->draftpost->groupid = $sqlconditions['groupid'];
+$SESSION->draftpost->courseid = $course->id;
+//rru-specific end
+
 echo $OUTPUT->footer();
 
